lib/coverage/enabled.ml

```diff
@@ -0,0 +1,117 @@
+(** This module sets up the env variable for bisect_ppx which describes
+ *  where log files are written.
+ *)
+
+module D=Debug.Make(struct let name="coverage" end)
+let prefix = "org.xen.xapi.coverage"
+
+module Bisect = struct
+  let bisect_file = "BISECT_FILE"
+  let dump jobid =
+    let bisect_prefix = Unix.getenv bisect_file in
+    (* dump coverage information in same location as it would normally
+       get dumped on exit, except also embed the jobid to make it easier to group.
+       Relies on [open_temp_file] generating a unique filename given a prefix/suffix to
+       avoid clashes with filenames created at exit by bisect itself. *)
+    let tmp, ch = Filename.open_temp_file
+                    ~temp_dir:(Filename.dirname bisect_prefix)
+                    (Filename.basename bisect_prefix)
+                    (Printf.sprintf ".%Ld.out" jobid) in
+    try
+      Bisect.Runtime.dump_counters_exn ch;
+      D.debug "Saved coverage data to %s" tmp;
+      close_out_noerr ch;
+      (* Keep file - will be collected by XenRT *)
+      tmp
+    with e ->
+      Sys.remove tmp;
+      D.warn "Failed to save coverage: %s" (Printexc.to_string e);
+      raise e
+
+  let reset () =
+    Bisect.Runtime.reset_counters ();
+    D.debug "Coverage counters reset"
+
+  let init_env name =
+    let (//)    = Filename.concat in
+    let tmpdir  = Filename.get_temp_dir_name () in
+    try
+      ignore (Sys.getenv bisect_file)
+    with Not_found ->
+      Unix.putenv bisect_file (tmpdir // Printf.sprintf "bisect-%s-" name)
+
+  let process body =
+    match Stringext.split ~on:' ' body with
+    | ["reset"] -> reset (); ""
+    | ["dump"; jobid] -> jobid |> Int64.of_string |> dump
+    | _ -> failwith body
+
+  let init name =
```

Comment:
What happens if `init` is called more than once? Is it worth defending against that?

Reply:
I think you'd have multiple threads competing to pull messages from the same message queue, but I don't see anything particularly bad happening, so I wouldn't protect it.